ckan/lib/changes.py

```diff
@@ -0,0 +1,771 @@
+        _title_change(change_list, original, new)
+
+    # if the owner organization changed
+    if original['owner_org'] != new['owner_org']:
+        _org_change(change_list, original, new, new_pkg)
+
+    # if the maintainer of the dataset changed
+    if original['maintainer'] != new['maintainer']:
+        _maintainer_change(change_list, original, new, new_pkg)
+
+    # if the maintainer email of the dataset changed
+    if original['maintainer_email'] != new['maintainer_email']:
+        _maintainer_email_change(change_list, original, new, new_pkg)
+
+    # if the author of the dataset changed
+    if original['author'] != new['author']:
+        _author_change(change_list, original, new, new_pkg)
+
+    # if the author email of the dataset changed
+    if original['author_email'] != new['author_email']:
+        _author_email_change(change_list, original, new, new_pkg)
+
+    # if the visibility of the dataset changed
+    if original['private'] != new['private']:
+        change_list.append([u"Set visibility of", new_pkg, u"to",
+                            u"Private" if new['private'] else u"Public"])
+
+    # if the description of the dataset changed
+    if original['notes'] != new['notes']:
+        _description_change(change_list, original, new, new_pkg)
+
+    # make sets out of the tags for each dataset
+    original_tags = set([tag['name'] for tag in original['tags']])
+    new_tags = set([tag['name'] for tag in new['tags']])
+    # if the tags have changed
+    if original_tags != new_tags:
+        _tag_change(change_list, new_tags, original_tags, new_pkg)
+
+    # if the license has changed
+    if original['license_title'] != new['license_title']:
+        _license_change(change_list, original, new, new_pkg)
+
+    # if the name of the dataset has changed
+    # this is only visible to the user via the dataset's URL,
+    # so display the change using that
+    if original['name'] != new['name']:
+        _name_change(change_list, original, new)
+
+    # if the source URL (metadata value, not the actual URL of the dataset)
+    # has changed
+    if original['url'] != new['url']:
+        _source_url_change(change_list, original, new, new_pkg)
+
+    # if the user-provided version has changed
+    if original['version'] != new['version']:
+        _version_change(change_list, original, new, new_pkg)
+
+    # check whether fields added by extensions or custom fields
+    # (in the "extras" field) have been changed
+
+    _extension_fields(change_list, original, new, new_pkg)
+    _extra_fields(change_list, original, new, new_pkg)
+
+
+def _title_change(change_list, original, new):
+    '''
+    Appends a summary of a change to a dataset's title between two versions
+    (original and new) to change_list.
+    '''
+    s = u""
+    seq2 = (u"<a href=\"", url_for(qualified=True, controller=u"dataset",
+            action=u"read", id=new['name']), u"\">",
+            new['title'], u"</a>")
```

Comment:
I guess there are some issues with building HTML in code:
* Firstly it is a security red-flag, putting user strings straight in. Using Jinja2 would escape any fragments of HTML. You could use `jinja2.Template(u'<a href="{{url}}">{{title}}</a>').render(url=x, title=y)` to fix that. However it would still suffer from:
* The strings are not translated (internationalization). You could add _('some string'), however the string would not be added to the ckan.pot file (i.e. 'message extraction' - created with `python setup.py extract_messages`). It normally picks up strings from .html files, although it would be a pain to add dozens of little .html files. Perhaps others (@smotornyuk @tino097 ) know a way to extract messages from code?

Reply:
I think I've fixed this with commit 7ccd68a12d100e6650225a52809efd6037bcd14d; let me know if that commit doesn't fix these issues.

Reply:
yep this is good!